File: BaekjoonOnlineJudge/1717/1717.py
import sys

input = sys.stdin.readline

def find(target):
    if target == disjoint_set[target]:
        return target
    
    disjoint_set[target] = find(disjoint_set[target])
    return disjoint_set[target]


def union(a, b):
    find_a = find(a)
    find_b = find(b)

    if find_a > find_b:
        disjoint_set[find_b] = find_a
    elif find_a < find_b:
        disjoint_set[find_a] = find_b
    else:
        return


n, m = map(int, input().split())

disjoint_set = list(range(n+1))

for _ in range(m):
    oper, a, b = map(int, input().split())

    if oper == 0:
        union(a, b)
    else:
        if find(a) == find(b):
            print("YES")
        else:
            print("NO")

File: BaekjoonOnlineJudge/1717/1717.java
import java.io.*;
import java.util.*;


public class Main {

    private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    private static StringTokenizer st;

    private static int[] disjointSet;

    private static int find(int target) {
        if (disjointSet[target] == target) {
            return target;
        }

        disjointSet[target] = find(disjointSet[target]);
        return disjointSet[target];
    }

    private static void union(int a, int b) {
        int findA = find(a);
        int findB = find(b);

        if (findA < findB) {
            disjointSet[findB] = findA;
        } else {
            disjointSet[findA] = findB;
        }
    }

    public static void main(String[] args) throws IOException {
        st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        disjointSet = new int[n + 1];

        for (int i = 1; i < n + 1; i++) {
            disjointSet[i] = i;
        }

        while (m-- != 0) {
            st = new StringTokenizer(br.readLine());

            int oper = Integer.parseInt(st.nextToken());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());

            if (oper == 0) {
                union(a, b);
            } else {
                if (find(a) == find(b)) {
                    bw.write("YES\n");
                } else {
                    bw.write("NO\n");
                }
            }
        }

        bw.flush();
        bw.close();
        br.close();
    }
}